### PR TITLE
Fixes #211 Improve the way memos are constructed

### LINKED ISSUE #211: Improve the way memos are constructed
Improve the way memos are handled. E.g., add a helper method to just add a plain-text memo as a String. See [here](https://github.com/sappenin/xrp-ping-pong/blob/main/src/main/java/com/sappenin/xrpl/pingpong/PingPong.java#L158) for an example.

### PR DESCRIPTION
**Pull request by sappenin**
* Add checks to make sure that memos Strings are hex-encoded
* Add a utility method to construct a memo with text/plain content-type (format)

Signed-off-by: David Fuelling <sappenin@gmail.com>

### CODE REVIEW

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java [hunk 1]
```diff
@@ -37,24 +48,66 @@
...
+   * Determines if an input char is hex-encoded.
+   *
+   * @param input A {@link char}.
+   *
+   * @return {@code true} if {@code input} is hex-encoded; {@code false} otherwise.
+   */
+
+  private static boolean isHex(final char input) {
+    return (input >= '0' && input <= '9') || (input >= 'a' && input <= 'f') || (input >= 'A' && input <= 'F');
+  }
```
Comment (nkramer44) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:112]: What happens here when you pass in a `String` with an odd number of characters, which is invalid hex?

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java [hunk 2]
```diff
@@ -20,7 +25,13 @@
...
 @JsonSerialize(as = ImmutableMemo.class)
 @JsonDeserialize(as = ImmutableMemo.class)
-public interface Memo {
+public abstract class Memo {
+
+  static ImmutableMemo.Builder withPlainText(final String plainText) {
+    return builder()
+      .memoFormat(BaseEncoding.base16().encode("text/plain".getBytes(StandardCharsets.UTF_8)))
+      .memoData(BaseEncoding.base16().encode(plainText.getBytes(StandardCharsets.UTF_8)));
+  }
```
Comment (nkramer44) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:34]: Should we also include a static constructor that lets you pass in a plaintext value for `MemoType`?

**Review by nkramer44 — CHANGES_REQUESTED** [review]

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java (hunk 2 above)
Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:34]: Not sure really - I didn't want to go overboard here except to support a plain-text memo. `MemoType` is meant to express a Link-Relation and I feel like anyone wanting to do that will need to encoded their URL as HEX manually -- this PR is really just meant to provide a simple utility to make plain-text easier, and to perhaps illustrate to devs how this should work.

Comment (nkramer44) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:34]: Not opposed to this for now. We can always add another constructor later on.

> I feel like anyone wanting to do that will need to encoded their URL as HEX manually 

Shouldn't we provide a utility for them to encode URLs to hex automatically? Is there something about hex encoding URLs that is user specific?

Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:34]: @nkramer44 and I decided to leave this out for now.

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java [hunk 3]
```diff
@@ -37,24 +48,66 @@
...
+
+  /**
+   * Determines if an input string is hex-encoded.
+   *
+   * @param input A {@link String}.
+   *
+   * @return {@code true} if every character in {@code input} is hex-encoded; {@code false} otherwise.
+   */
+  private static boolean isHex(final String input) {
+    Objects.requireNonNull(input);
```
Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:92]: Try Guava

Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java:99]: rippled supports nibbles in a memo, whereas Guava does not. So this seems correct as-is.

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java (hunk 1 above)
Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Memo.java]: Confirmed that this works with rippled in an IT.

**Review by nkramer44 — APPROVED** [review]

**Pull request merged**